### DISCUSSION TITLE
pulsar standalone docker image should support initialization of some kind

### DISCUSSION THREAD

**Discussion by flowchartsman** [apache/pulsar#18781] [General]
This would be very useful for development and integration testing. Currently have to run a parallel container calling `pulsar-admin` in a loop to wait on the relevant endpoints to be up and spin up the tenants/namespaces/policies that I need to simulate my deployment environment.

Perhaps this could be read from a file at a known location and then spun up in place fo the sample tenant/namespaces if it's found?

**Comment by flowchartsman** [comment]
My suggestion would be something like the following:
1) before spinning up the admin endpoint, look for a file in a known location (following mongo's example, this would be something like `/docker-entrypoint-init.d/pulsar-init`
2) If this file is found, process it
3) otherwise, proceed with `sample` tenant/namespace creation

As for the format of the file, one option would be to simply treat it as a newline-separated list of arguments to `pulsar-admin` invocations:

```
tenants create test
namespaces create test/admin
namespaces create test/news
namespaces create test/emails
namespaces set-retention --size-1 --time 3d test/news
namespaces set-publish-rate -m 10 test/emails
```

**Comment by tisonkun** [comment]
Moved to the General Discussions forum. I like the idea but it requires a design (PIP) and volunteers working on it. Discussions forum is sorted default by updated and gives this thread more attention.

Perhaps `.hiverc` to Hive. For your suggestion command pattern, the problem is which bin script to execute, shall we support run `bin/pulsar ...` also?

**Reply by flowchartsman** [reply]
> shall we support run bin/pulsar ... also?

Not for the use-case I had in mind, which was for pulsar, specifically in standalone mode. Actually, even more specifically, it was targeted at the `pulsar-standalone` image to help out in development and testing.  Since I originally wrote this issue, it seems the dedicated `pulsar-standalone` image has fallen behind, so it makes it tricker to implement, since you'd either need to have a dedicated image, or build it into pulsar directly, and that could be problematic if misused.

Since then, I've also found a technique that works "well enough", by using a script which calls curl on the metrics endpoint in a loop,  then setting this script as the as the container "health check":

```bash
#!/bin/bash
cd /pulsar/bin
# Wait for pulsar standalone to be initialized
until curl http://localhost:8080/metrics > /dev/null 2>&1 ; do sleep 1; done
./pulsar-admin tenants create -c standalone mytenant
./pulsar-admin namespaces create mytenant/mynamespace
<etc>
```

So I guess my question at this point would be: is it useful to provide this functionality as described, or is there a compelling use-case for calling `/pulsar/bin/pulsar` as well? I'm not sure, by I still think this would be handy, even if it's config-file driven.
